src/goto-programs/goto_contractor.cpp

```diff
@@ -0,0 +1,453 @@
+//
+// Created by Mohannad Aldughaim on 09/01/2022.
+//
+
+#include "goto_contractor.h"
+
+void goto_contractort::get_constraints(goto_functionst goto_functions)
+{
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assert())
+      constraint = create_constraint_from_expr2t(it->guard);
+
+    it++;
+  }
+}
+
+void goto_contractort::get_intervals(goto_functionst goto_functions)
+{
+  //only main here
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assume())
+    {
+      it->dump();
+      parse_intervals(it->guard);
+    }
+
+    it++;
+  }
+}
+
+void goto_contractort::parse_intervals(irep_container<expr2t> expr)
+{
+  symbol2tc symbol;
+  long value;
+
+  expr = get_base_object(expr);
+
+  if(is_and2t(expr))
+  {
+    parse_intervals(to_and2t(expr).side_1);
+    parse_intervals(to_and2t(expr).side_2);
+    return;
+  }
+
+  if(!is_comp_expr(expr))
+    return;
+
+  std::shared_ptr<relation_data> rel;
+  rel = std::dynamic_pointer_cast<relation_data>(expr);
+
+  //side1 1 is a symbol or typecast to symbol
+  auto side1 = rel->side_1;
+  //side 2 is always a number.
+  auto side2 = rel->side_2;
+
+  auto obj = get_base_object(side1);
+  if(is_symbol2t(obj))
+    symbol = to_symbol2t(obj);
+  else
+    return;
+  ///keeping this for later
+  //  if(is_typecast2t(side1))
+  //    if(is_symbol2t(to_typecast2t(side1).from))
+  //      symbol = to_symbol2t(to_typecast2t(side1).from);
+  //    else
+  //      return;
+  //  else if(is_symbol2t(side1))
+  //    symbol = to_symbol2t(side1);
+  //  else
+  //    return;
+
+  //message_handler.status(get_expr_id(side2));
+  //side2->dump();
+  bool neg = false;
+  if(is_neg2t(side2))
+  {
+    neg = true;
+    side2 = to_neg2t(side2).value;
+  }
+  if(!is_constant_int2t(side2) && !is_constant_floatbv2t(side2))
+    return;
+  if(is_constant_int2t(side2))
+    value = to_constant_int2t(side2).as_long() * (neg?-1:1);
+  else
+    return;
+
+  int index = map->find(symbol->get_symbol_name());
+  if(index == NOT_FOUND)
+    return;
```

Comment:
what's not_found here?

Reply:
It is defined to indicate that this symbol is not in the tracked symbols list. Its value is -1.